components/constellation/session_history.rs

```diff
@@ -0,0 +1,160 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use msg::constellation_msg::{BrowsingContextId, PipelineId, TopLevelBrowsingContextId};
+use script_traits::LoadData;
+use std::{fmt, mem};
+use std::cmp::PartialEq;
+
+/// Represents the joint session history
+/// https://html.spec.whatwg.org/multipage/#joint-session-history
+#[derive(Debug)]
+pub struct SessionHistory {
+    /// Diffs used to traverse to past entries
+    pub past: Vec<SessionHistoryDiff>,
+
+    /// Diffs used to traverse to future entries
+    pub future: Vec<SessionHistoryDiff>,
+}
+
+impl SessionHistory {
+    pub fn new() -> SessionHistory {
+        SessionHistory {
+            past: Vec::new(),
+            future: Vec::new(),
+        }
+    }
+
+    pub fn history_length(&self) -> usize {
+        self.past.len() + 1 + self.future.len()
+    }
+
+    pub fn push_diff(&mut self, diff: SessionHistoryDiff) -> Vec<SessionHistoryDiff> {
+        self.past.push(diff);
+        mem::replace(&mut self.future, vec![])
+    }
+
+    pub fn replace(&mut self, old_pipeline_id: AliveOrDeadPipeline, new_pipeline_id: AliveOrDeadPipeline) {
+        for diff in self.past.iter_mut().chain(self.future.iter_mut()) {
+            diff.replace(&old_pipeline_id, &new_pipeline_id);
+        }
+    }
+}
+
+/// Represents a pending change in a session history, that will be applied
+/// once the new pipeline has loaded and completed initial layout / paint.
+pub struct SessionHistoryChange {
+    /// The browsing context to change.
+    pub browsing_context_id: BrowsingContextId,
+
+    /// The top-level browsing context ancestor.
+    pub top_level_browsing_context_id: TopLevelBrowsingContextId,
+
+    /// The pipeline for the document being loaded.
+    pub new_pipeline_id: PipelineId,
+
+    /// The old pipeline that the new pipeline should replace.
+    pub replace: Option<AliveOrDeadPipeline>,
+}
+
+/// Represents a pipeline or discarded pipeline in a history entry.
+#[derive(Clone, Debug)]
+pub enum AliveOrDeadPipeline {
+    /// Represents a pipeline that has not been discarded
+    Alive(PipelineId),
+    /// Represents a pipeline that has been discarded and must be reloaded with the given `LoadData`
```

Comment:
Nit: "must be reloaded with the given `LoadData` if it is ever traversed to".

Reply:
Done.